### PR TITLE
Hyper Parameters

### PR DESCRIPTION
**Pull request by elephantmipt**
## Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contribution guide](https://github.com/catalyst-team/catalyst/blob/master/CONTRIBUTING.md)?
- [ ] Did you check the code style? `catalyst-make-codestyle && catalyst-check-codestyle` (`pip install -U catalyst-codestyle`).
- [ ] Did you make sure to update the docs? We use Google format for all the methods and classes.
- [ ] Did you check the docs with `make check-docs`?
- [ ] Did you write any new necessary tests?
- [ ] Did you add your new functionality to the docs?
- [ ] Did you update the [CHANGELOG](https://github.com/catalyst-team/catalyst/CHANGELOG.md)?
- [x] **You can use 'Login as guest' to see Teamcity build logs.**

## Type of Change

- [ ] Examples / docs / tutorials / contributors update
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves an existing feature)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

### CODE REVIEW

File: catalyst/dl/experiment/config.py [hunk 1]
```diff
@@ -80,6 +80,11 @@ def __init__(self, config: Dict):
             self._config["stages"]
         )
 
+    @property
+    def hparams(self):
```
Comment (Scitator) [inline catalyst/dl/experiment/config.py]: could you please add here ` -> OrderedDict:` too?

File: catalyst/dl/experiment/experiment.py [hunk 2]
```diff
@@ -143,6 +143,30 @@ def distributed_params(self) -> Dict:
...
+        """Returns hyper parameters"""
+        hparams = OrderedDict()
+        if self._optimizer is not None:
+            optimizer = self._optimizer
+            hparams["optimizer"] = optimizer.__repr__().split()[0]
+            params_dict = optimizer.state_dict()["param_groups"][0]
+            for k, v in params_dict.items():
+                if k != "params":
+                    hparams[k] = v
+        loaders = self.get_loaders("train")
```
Comment (Scitator) [inline catalyst/dl/experiment/experiment.py]: `self._stage` maybe?

File: catalyst/dl/experiment/experiment.py [hunk 3]
```diff
@@ -143,6 +143,30 @@ def distributed_params(self) -> Dict:
...
+        num_train_loaders = 0
+        for k in loaders.keys():
+            if "train" in k:
```
Comment (Scitator) [inline catalyst/dl/experiment/experiment.py]: it's better to use `.startswith()`

File: catalyst/dl/experiment/experiment.py [hunk 4]
```diff
@@ -143,6 +143,30 @@ def distributed_params(self) -> Dict:
...
+                num_train_loaders += 1
+        if num_train_loaders == 1:
+            hparams["batch_size"] = loaders["train"].batch_size
+        else:
+            for k, v in loaders.items():
+                if "train" in k:
+                    hparams[f"{k}_batch_size"] = v.batch_size
```
Comment (Scitator) [inline catalyst/dl/experiment/experiment.py]: I think, it's better to use the same naming like this for all experiments, even if we have only 1 loader

File: catalyst/dl/experiment/experiment.py [hunk 5]
```diff
@@ -143,6 +143,30 @@ def distributed_params(self) -> Dict:
         """Dict with the parameters for distributed and FP16 method."""
         return self._distributed_params
 
+    @property
+    def hparams(self) -> OrderedDict:
+        """Returns hyper parameters"""
```
Comment (Scitator) [inline catalyst/dl/experiment/experiment.py:148]: btw, could you please also add an example output here?
like https://github.com/catalyst-team/catalyst/blob/24f147f52a25612a88b1e94df7a89ee31619caa4/catalyst/core/experiment.py#L77#L78

**Comment by elephantmipt** [comment]
I think `hparams` property should be the same in Notebook and Config API, so we can use the same code in all logging callbacks. Also hparams in config experiment is now feels like overkill because `experiment._config` and `experiment.hparams` returns same value.

**Comment by Scitator** [comment]
@PUSSYMIPT Notebook and Config API should provide the same API to use so we could log them into our monitoring tools easily, you are right. Nevertheless, the ultimate power of Config API is all hyperparameters logging through yml-config. You can easily filter unnecessary params during visualization (we are working on such feature for Alchemy), but you should log everything for reproducibility and insights findings.

**Comment by Scitator** [comment]
I will revisit this PR after https://github.com/catalyst-team/catalyst/pull/837 merge, so we could have perfect docs for everyone :)

**Review by Scitator — APPROVED** [review]

**Review by Scitator — DISMISSED** [review]
could you please add tests and changelog?

**Pull request merged**